diff --git a/src/aplt.sml b/src/aplt.sml
index 2a7db7f..4ac39ca 100644
--- a/src/aplt.sml
+++ b/src/aplt.sml
@@ -152,6 +152,6 @@ fun usage() =
     " -opt_loopsplit : enable loop split optimization in LAILA code generation\n"
 
 (* Parse command line arguments and pass to compileAndRun *)
-val () = Flags.runargs {usage = usage,
+fun main () = Flags.runargs {usage = usage,
                         run = compileAndRun,
                         unaries = ["-o","-oc","-O"]}
